stl/inc/iterator

```diff
@@ -513,6 +513,501 @@ private:
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return *_Val._Iterator;
+    }
+
+    // clang-format off
+    _NODISCARD decltype(auto) operator->() const
+        requires indirectly_readable<const _Iter>
+            && (_Has_op_arrow<_Iter> || is_reference_v<iter_reference_t<_Iter>>
+                || constructible_from<iter_value_t<_Iter>, iter_reference_t<_Iter>>) {
+        // clang-format on
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Val._Contains == _Vty::_State::_Holds_iter,
+            "common_iterator can only be dereferenced if it holds an iterator");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if constexpr (is_pointer_v<_Iter> || _Has_op_arrow<_Iter>) {
+            return (_Val._Iterator);
+        } else if constexpr (is_reference_v<iter_reference_t<_Iter>>) {
+            auto&& _Tmp = *_Val._Iterator;
+            return _STD addressof(_Tmp);
+        } else {
+            return _Proxy{*_Val._Iterator};
+        }
+    }
+
+    common_iterator& operator++() {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Val._Contains == _Vty::_State::_Holds_iter,
+            "common_iterator can only be incremented if it holds an iterator");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        ++_Val._Iterator;
+        return *this;
+    }
+
+    decltype(auto) operator++(int) {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Val._Contains == _Vty::_State::_Holds_iter,
+            "common_iterator can only be incremented if it holds an iterator");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if constexpr (forward_iterator<_Iter>) {
+            common_iterator _Tmp = *this;
+            ++_Val._Iterator;
+            return _Tmp;
+        } else {
+            return _Val._Iterator++;
+        }
+    }
+
+    // clang-format off
+    template <class _Iter2, sentinel_for<_Iter> _Se2>
+        requires sentinel_for<_Se, _Iter2>
+    _NODISCARD friend bool operator==(const common_iterator& _Left, const common_iterator<_Iter2, _Se2>& _Right) {
+        // clang-format on
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Left._Val._Contains != _Vty::_State::_Nothing && _Right._Val._Contains != _Vty::_State::_Nothing,
+            "common_iterators can only be compared if both hold a value");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+
+        if (_Left._Val._Contains == _Right._Val._Contains) {
+            return true;
+        }
+
+        if (_Left._Val._Contains == _Vty::_State::_Holds_iter) {
+            return _Left._Val._Iterator == _Right._Val._Sentinel;
+        } else {
+            return _Left._Val._Sentinel == _Right._Val._Iterator;
+        }
+    }
+
+    // clang-format off
+    template <class _Iter2, sentinel_for<_Iter> _Se2>
+        requires sentinel_for<_Se, _Iter2> && equality_comparable_with<_Iter, _Iter2>
+    _NODISCARD friend bool operator==(const common_iterator& _Left, const common_iterator<_Iter2, _Se2>& _Right) {

```

Comment:
We should merge this overload with the previous overload and use `if constexpr (equality_comparable_with<_Iter, _Iter2>)` to determine when to implement its behavior (subsumption is expensive).

Reply:
This makes it even simpler. Thanks!

Reply:
One of the major wording overhauls I made between the Ranges TS and C++20 was to refactor concept overloading into implicit `if constexpr` dispatch because it's easier to understand, easier to implement, and compiles faster. There are some places like this that I missed: keep your eyes peeled for overloads whose signatures are identical except for constraints, they can generally be `if constexpr`ed together easily unless conditional `noexcept` is involved. `noexcept` and `if constexpr` are natural enemies =(